cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaClusterRebalanceAssemblyOperator.java

```diff
@@ -373,6 +383,9 @@ public void onClose(KubernetesClientException e) {
                         // and the previous execution set the status and completed the future
                         if (state(freshClusterRebalance) == State.PendingProposal) {
                             if (getRebalanceAnnotation(freshClusterRebalance) == RebalanceAnnotation.stop) {
+                                // Question: If this is a pending proposal there is nothing for the CC REST API to stop.
+                                // The Stop Execution endpoint only stops active cluster altering tasks
+                                // https://github.com/linkedin/cruise-control/wiki/REST-APIs#stop-the-current-proposal-execution-task
```

Comment:
Being in a `PendingProposal` state means that the operator is waiting for the CC to provide a rebalance proposal after starting a `rebalance` operation with dryrun. As far as I can see from the link, it should be something stoppable, or?
Having a proposal ready, waiting for the user to be approved (`approve` annotation), happens in the `ProposalReady` state and in that case there is nothing to stop of course.

Reply:
Well, for that endpoint is says:

> The following POST request will let Kafka Cruise Control stop an ongoing rebalance, add_broker, remove_broker, fix_offline_replica, topic_configuration or demote_broker operation

That list does not include the calculation of a proposal but I will have to check the CC source code to be sure.

Reply:
Isn't it the `rebalance` endpoint? This is the one I am using for asking a rebalance proposal.

Reply:
No, in this case if the stop annotation is present you are calling the ` apiClient.stopExecution` which is a different endpoint to the `rebalance`. As far as I can see there is no way to stop the calculation of a "proposal" once it has started.

Reply:
What I meat was that the doc says:

> the following POST request will let Kafka Cruise Control stop an ongoing rebalance...

so it sounds calling `apiClient.stopExecution` should be able to stop an operation started on the `rebalance` endpoint which is the one I use for rebalance proposal.

Reply:
I think the we need to make clear the difference between a `rebalance(dryrun=true,customGoals=[goal1, goal2, etc])` which will start the calculation of a "proposal" and doing the above with `dryrun=true` which will start the actual data movements which the `stopExecution` endpoint will stop.

Reply:
What I understand from the doc is that the `rebalance` endpoint is about triggering a workload balance. When dryrun=true, it provides the proposal for the list of goals you specify. 
Then, as far as I can see it says:

> To let Kafka Cruise Control actually move data, users need to explicitly set dryrun=false

So it means that calling the `rebalance` endpoint with dryrun=false does the actual data movement.

Then, if the doc is right, it seems that the stop execution acts on the `rebalance` endpoint and it's not related to the fact you call it with dryrun=false|true, so it should be able to stop an ongoing proposal processing or actual movement.
Of course, I could have misunderstood the doc :-)

Reply:
Yes that is all true but this is in the `pendingProposal` method so there is currently no rebalance to stop. So when you are calling `stopExecution` in this method AFAIU it will do nothing?

Reply:
The `onPendingProposal` method is called when in `PendingProposal` state which means a `rebalance` endpoint was called with dryrun=true and we are waiting for the proposal to be ready so I guess it could be stopped by `stopExecution`.
To be clear, when a `KafkaClusterRebalance` resource is created we are in `New` state and the `onNew` method is called which calls `requestRebalance` which calls the `rebalance` endpoint with dryrun=true. At this point we can get a proposal already ready so we move to `ProposalReady` or, if the proposal is not ready, we move to `PendingProposal`. At this point, it's in the `onPendingProposal` that we check the status or eventually call `stopExecution` so while a rebalance proposal processing is ongoing

Reply:
Agreed, but calling `stopExecution` before reabalance(dryrun=false) won't have any effect as far as I understand it. But I am going over the CC code to make sure.

Reply:
> stopExecution before reabalance(dryrun=false)

But how can this happens? the stopExectuion is called only in the `PendingProposal` state which means you have already called `rebalance` with dryrun=false.

Reply:
Right, but the `PendingProposal` state represents:
```
       /**
         * A proposal has been requested from Cruise Control, but is not ready yet.
         * Transitions to:
         * <dl>
         *     <dt>ProposalReady</dt><dd>Once Cruise Control has a proposal.</dd>
         *     <dt>NotReady</dt><dd>If Cruise Control returned an error</dd>
         *     <dt>Stopped</dt><dd>If the user sets annotation strimzi.io/rebalance=stop.</dd>
         * </dl>
         */
         PendingProposal,
```
The "proposal" is requested using `dryrun=true`. So we are waiting for CC to tell us what it intends to do. From there we can go to `ProposalReady` and get on with getting approval from the user. Or we can call the whole thing off by the user attaching "stop" to the KafkaRebalance CR. If we are stopping a pending "proposal" there is nothing to "stop" on the CC side, as the user has not approved the proposal so no data is moving, so there is no need to call the stopExecution endpoint? 

My main point is that:
```
if (state(freshClusterRebalance) == State.PendingProposal) {
    if (getRebalanceAnnotation(freshClusterRebalance) == RebalanceAnnotation.stop) {
         apiClient.stopExecution(........)
    }
}
```
Is not needed as a `stopExecution` call when there is no rebalance operation running (PendingProposal) will not actually do anything.

In that particular case we just attach the "stopped" status/annotation and move on?

Reply:
> If we are stopping a pending "proposal" there is nothing to "stop" on the CC side, as the user has not approved the proposal so no data is moving.

My understanding is that you can stop the running process of CC to find a proposal. It's not actually moving data of course but it's doing some processing to find a proposal for you. Isn't it stoppable?